### PR TITLE
[MRG+1] Several hatching improvements

### LINKED ISSUE #3023: contourf hatching and saving to pdf
`contourf` hatches do not appear when saving to pdf. This is apparent in the [`contourf_hatching.py`](http://matplotlib.org/examples/pylab_examples/contourf_hatching.html) example, here:

http://matplotlib.org/mpl_examples/pylab_examples/contourf_hatching_01.pdf

I think something similar was mentioned in #891 and #29, but I'm not sure it was resolved.

### LINKED ISSUE #4108: Hatch pattern changes with dpi
Hi, I currently use matplotlib a lot - thanks for your nice work! I noticed that the hatch patterns change with the dpi settings. This is not a big problem as long as you raster the whole plot. However, if you only raster parts of it, you will find that the **legend not longer matches the pattern** in the bars and that there is a **gap at the bottom** of the bars.

```
import matplotlib.pyplot as plt
fig = plt.gcf()
ax = fig.add_subplot(111)
ax.bar(0.7, 5, 0.3, label="raster /", hatch="////", rasterized=True)
ax.bar(1.0, 7, 0.3, label="vector /", hatch="////", rasterized=False)
ax.bar(1.7, 5, 0.3, label="raster .", hatch="....", rasterized=True)
ax.bar(2.0, 7, 0.3, label="vector .", hatch="....", rasterized=False)
ax.bar(2.7, 5, 0.3, label="raster x", hatch="xxxx", rasterized=True)
ax.bar(3.0, 7, 0.3, label="vector x", hatch="xxxx", rasterized=False)
ax.set_ylim([0,10])
plt.tight_layout()
ax.legend(loc=0,ncol=3)
fig.savefig("dpi050.png", dpi=50)
fig.savefig("dpi300.png", dpi=300)
fig.savefig("dpi050.pdf", dpi=50)
fig.savefig("dpi300.pdf", dpi=300)
```

And this is how `dpi050.pdf` looks:
dpi050

### LINKED ISSUE #6228: Rasterizing patch changes filling of hatches in pdf backend
I want to plot a patch with the '*' hatch that is filled. The hatch is always filled when using
the agg backend but in the pdf backend only when it is rasterized. 

As the legend artists do not use the rasterized prop (correct?), it is inconsistent.

mpl: 1.5.1, python 3.5.1, arch linux

Code to reproduce:

```
import matplotlib
import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot(111)

# patch without hatching
patch1 = matplotlib.patches.Rectangle((0.5, 0.5), 1., 1., color='red', fill=False, hatch='*', label='patch1')
# patch with hatching
patch2 = matplotlib.patches.Rectangle((2.0, 0.5), 1., 1., color='blue', fill=False, hatch='*', rasterized=True, label='patch2')

ax.add_patch(patch1)
ax.add_patch(patch2)

ax.set_xlim(0., 3.5)
ax.set_ylim(0., 3)

ax.legend()


fig.savefig('asdf.png')
fig.savefig('asdf.pdf')
```

PNG output:

asdf

PDF output:

asdf pdf

### PR DESCRIPTION
**Pull request by QuLogic**
This produces some more consistency for hatches across multiple backends.
 * Get Agg to produce DPI-independent hatching; fixes #4108.
 * Fix PostScript hatching with edgeless artists.
 * Make hatching origin consistent with Agg.
 * Fill hatches on other backends to match Agg; fixes #3023; fixes #6228.

### CODE REVIEW

**Comment by tacaswell** [comment]
Would be good if @mdboom could look at this, but :+1: 

Holding off on merging this until @QuLogic has looked at the hatch origin issue seems reasonable to me to avoid having to re-generate the test images again.

**Comment by QuLogic** [comment]
I've fixed the hatching origin in the PDF and PS backends, and also made a few other improvements in the PS backend. See the description for an updated explanation.

Note, a lot of the PS and PDF backend doesn't really seem to be tested... Could add the test based on the related issues, but not sure where the best place is.

**Comment by NelleV** [comment]
Thanks for the patch!

**Comment by jkokorian** [comment]
How can I set the hatch line color when using `fill_between`? I am only able to set the background color using the `color` kwarg. `edgecolor` doesn't appear to do anything...

**Comment by tacaswell** [comment]
@jkokorian See #7750

**Comment by tacaswell** [comment]
@jkokorian sorry meant #7976

**Comment by jkokorian** [comment]
thanks!

**Comment by andres0l1** [comment]
Hi, Sorry I am a bit new to Python but I am having the same problems as in: [#3023 ](https://github.com/matplotlib/matplotlib/issues/3023/) I have re-install Matplotlib and updated it but my hatches still do not show up when I save my plot as .pdf
Thanks!

**Comment by tacaswell** [comment]
@andres0l1 Can you please open a new issue with a minimal example of what is not working for you?

**Comment by andres0l1** [comment]
Sure I will!

File: lib/matplotlib/backends/backend_ps.py [hunk 1]
```diff
@@ -322,6 +322,7 @@ def create_hatch(self, hatch):
             self._convert_path(Path.hatch(hatch), Affine2D().scale(sidelen),
                                simplify=False))
         self._pswriter.write("""\
+        fill
```
[the remark this answers is not in the document]
Comment (mksegv) [reply lib/matplotlib/backends/backend_ps.py]: Or shell we use the `shfill` command?

[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/backends/backend_ps.py]: What is the difference between `fill` and `shfill`?

[the remark this answers is not in the document]
Comment (mksegv) [reply lib/matplotlib/backends/backend_ps.py]: @tacaswell sorry, I just skimmed [the spec of EPS](https://www.adobe.com/content/dam/acom/en/devnet/actionscript/articles/5002.EPSF_Spec.pdf) and found that `shfill` (shading fill, which "paints the shape and color shading described by a shading dictionary, subject to the current clipping path") seems not related.

But I do have tried multiple manual fixes on the generated eps file and had the following findings:
1. Simply removing the "fill" can solve the issue in #8289. But when we draw with "*", the stars are not filled with colors, which breaks #6228.
2. Surrounding the `fill` with `gsave` and `grestore`, also fixes #8289, but the result of #6228 is as follow.

image

I'm not familiar with the eps format, but hope the information is helpful.

**Pull request merged**